x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointAction.java

```diff
@@ -0,0 +1,235 @@
+            return new Response(seqNoStats.getGlobalCheckpoint());
+        }
+
+        @Override
+        protected void asyncShardOperation(Request request, ShardId shardId, ActionListener<Response> listener) throws IOException {
+            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+            final SeqNoStats seqNoStats = indexShard.seqNoStats();
+
+            if (request.waitForAdvance() && request.currentCheckpoint() >= seqNoStats.getGlobalCheckpoint()) {
+                indexShard.addGlobalCheckpointListener(
+                    request.currentCheckpoint() + 1,
+                    new GlobalCheckpointListeners.GlobalCheckpointListener() {
+
+                        @Override
+                        public Executor executor() {
+                            return threadPool.executor(ThreadPool.Names.GENERIC);
+                        }
+
+                        @Override
+                        public void accept(final long g, final Exception e) {
+                            if (g != UNASSIGNED_SEQ_NO) {
+                                assert request.currentCheckpoint() <= g
+                                    : shardId + " only advanced to [" + g + "] while waiting for [" + request.currentCheckpoint() + "]";
+                                globalCheckpointAdvanced(shardId, request, listener);
+                            } else {
+                                assert e != null;
+                                globalCheckpointAdvancementFailure(shardId, e, listener);
+                            }
+                        }
+
+                    },
+                    request.pollTimeout());
+            } else {
+                super.asyncShardOperation(request, shardId, listener);
+            }
+        }
+
+        private void globalCheckpointAdvanced(
+            final ShardId shardId,
+            final Request request,
+            final ActionListener<Response> listener) {
+            try {
+                super.asyncShardOperation(request, shardId, listener);
+            } catch (final IOException caught) {
+                listener.onFailure(caught);
+            }
+        }
+
+        private void globalCheckpointAdvancementFailure(final ShardId shardId, final Exception e, final ActionListener<Response> listener) {
+            if (e instanceof TimeoutException) {
+                try {
+                    final IndexMetadata indexMetadata = clusterService.state().metadata().index(shardId.getIndex());
+                    // TODO: Necessary?
+                    if (indexMetadata == null) {
+                        listener.onFailure(new IndexNotFoundException(shardId.getIndex()));
+                    } else {
+                        listener.onFailure(e);
+                    }
+                } catch (final Exception caught) {
```

Comment:
Can we avoid this? We risk double firing listeners if `listener.onFailure` fails.